오픈마켓 I [STEP 1] Donnie, 우롱차

westeastyear opened the pull request:
안녕하세요!! @TTOzzi 🙇🏻‍♂️🙇🏻‍♂️
도니, 우롱차입니다.
첫번째 PR 보냅니다.
저희 회의실은 언제나 활짝 열려있습니다. 잘 부탁드리겠습니다!!! 😘

## 🧐 고민한 점 및 해결한 부분

#### 1. `requestData` 구현시 어디까지 `escaping Handler`에서 처리할것이고, 어디까지 함수 내부에서 처리를 한 뒤 `Handler`에게 넘겨줄지를 고민하였습니다.
- 의논 결과 전부 다 `Handler`에서 처리하도록 하자고하여 `data`, `response`, `error`를 전부 다 넘겼습니다.
```swift
func requestData(url: String, completeHandler: @escaping (Data?, URLResponse?, Error?) -> Void ) {
        let configure = URLSessionConfiguration.default
        let session = URLSession(configuration: configure)
        
        let urlComponents = URLComponents(string: url)
        
        guard let requestURL = urlComponents?.url else { return }
        let dataTask = session.dataTask(with: requestURL) { (data, response, error) in
            guard error == nil else { return }
            completeHandler(data, response, error)
        }
```

#### 2. `URL` 주소를 관리하는 방법에 대해
- 처음엔 함수에서 `query` 매개변수를 받아 `URLQueryItem`으로 `URL`를 관리하는 방법을 채택했었는데, 어떤 `query` 파라미터를 사용해야 하는지 알기 어려워져서 다른 방법을 사용하게 되었습니다.
*(아무 String 값이 들어갈 수 있어서 새로 작성하는 입장에서는 어떤 값이 들어가는지 모르고 오류가 나기 쉽기 때문에 `enum`으로 관리하도록 변경하였습니다.)*
```swift
func inquireProductList(url: String, query: [(String, String)], completeHandler: @escaping (Data?, URLResponse?, Error?) -> Void ) {
    
------중략------
    
    var urlComponents = URLComponents(string: url)
    for item in query {
        let urlQueryItem = URLQueryItem(name: item.0, value: item.1)
        urlComponents?.queryItems?.append(urlQueryItem)
    }
}
```

- 팀원과 `URL` 주소를 관리하는 법에 대하여 고민하였고, `enum`으로 관리해주기로 결정하였습니다.
```swift
enum OpenMarketApi {
    private static let hostUrl = "https://market-training.yagom-academy.kr/"
    
    case pageInformation(pageNo: Int, itemsPerPage: Int)
    case productDetail(productNumber: Int)
    
    var string: String {
        switch self {
        case .pageInformation(let pageNo, let itemsPerPage):
            return Self.hostUrl + "api/products?page_no=\(pageNo)&items_per_page=\(itemsPerPage)"
        case .productDetail(let productNumber):
            return Self.hostUrl + "api/products/\(productNumber)"
        }
    }
}
```

<br>

## 😆 조언을 받고 싶은 부분
#### 1. 네이밍 `Network`가 부적절한거 같은데 딱히 더 좋은 방안이 생각나지 않았습니다. 네이밍에 대한 많은 조언 부탁드립니다.🙇🏻‍♂️🙇🏻‍♂️

<br>

#### 2. 테스트하는 `JSON`파일이 `main asset`에 들어가 있는게 좋아보이지 않아서 `test`파일에 위치해두어 `testBundle`로 접근하게 하였습니다. 저희는 괜찮다고 생각이 드는데, 또치의 생각은 어떠한가요? 
<img src="https://user-images.githubusercontent.com/74251593/167859277-e1c41135-a68e-4d8a-b88b-b2c90fee4d61.png" width="60%">

#### 3. `Session`으로부터 `response`를 받을때 모든 `data`를 저장하도록 객체를 설계하였는데, 필요한 `data`만 디코딩하여 저장하는것이 나은지 궁금하여 조언을 받고싶습니다.
```swift
//모든 data를 저장하는 경우
struct PageInformation: Codable {
    let pageNo: Int
    let itemsPerPage: Int
    let totalCount: Int
    let offset: Int
    let limit: Int
    let pages: [ProductInformation]
    let lastPage: Int
    let hasNext: Bool
    let hasPrev: Bool
}
```

```swift
//불필요한 data를 제거하고 저장하는 경우
struct PageInformation: Codable {
    let pageNo: Int
    let itemsPerPage: Int
    let totalCount: Int
    let pages: [ProductInformation]
    let lastPage: Int
}
```

<br>

## 🥄 삽질한점 🥄

#### 1. `decoding`시 `CodingKeys` 쪽에 오타가 많아 이를 해결하는 과정이 오래 걸렸습니다.🥲🥲 (ex: `Codingkeys` -> `CodingKeys`)

<br>

dnwhd0112 committed:
📝[docs]: README 수정

README 수정

dnwhd0112 committed:
Update README.md

dnwhd0112 committed:
✨[feat]: JsonParser 파일 추가

dnwhd0112 committed:
✅[chore]: 작성자 이름 변경

dnwhd0112 committed:
✨[feat]: PageInformation struct 데이터 모델 구현

westeastyear committed:
✨[feat]: decodingJson(), encodingJson() 메서드 구현

dnwhd0112 committed:
✨[feat]: 파일 추가, Currency enum, productInformation struct 타입 구현

westeastyear committed:
✅[chore]: 가격 관련 상수 Double 타입으로 변경

westeastyear committed:
✨[feat]: ProductDetail, Image, Vendor 모델 타입 구현

dnwhd0112 committed:
➕[add]: Codable 채택 추가

dnwhd0112 committed:
✨[feat]: RegistrationParameter struct, ProductRegistration struct 구현

westeastyear committed:
✨[feat]: ProductDeleteInformation, ProductSecret 모델타입 구현

westeastyear committed:
✨[feat]: 테스트 파일 구현

westeastyear committed:
✅[chore]: 오타 수정 및 테스트 코드 작동 확인

dnwhd0112 committed:
✨[feat]: 테스트 코드 수정

westeastyear committed:
⚰️[del]: 쓸모없는 코드 삭제.

westeastyear committed:
✨[feat]: NetworkAble 프로토콜 정의 및 MockTestData 객체 구현

dnwhd0112 committed:
✨[feat]: NetworkAble 수정 및 테스트 코드 수정

westeastyear committed:
✨[feat]: productDetail Decoding 테스트 함수 구현 및 리펙토링

DataType, Network으로 그룹 분리
OpenMarketApiUrl 타입 생성
NetworkAble query 파라미터 옵셔널로 변경

dnwhd0112 committed:
✅[chore]: 에러 처리 추가

dnwhd0112 committed:
♻️[refactor]: url 얻는 방식 수정

westeastyear committed:
Merge branch 'STEP1-2'

dnwhd0112 committed:
♻️[refactor]: URL 관리 방식 수정 및 Session task 시 Query 불필요하도록 변경

westeastyear committed:
✏️[correct]: 파일, 변수, 상수명 변경

westeastyear committed:
Merge branch 'STEP1-2'

westeastyear committed:
⚰️[del]: 쓸모없는 error 코드 제거

westeastyear committed:
Merge branch 'STEP1-2'

TTOzzi commented on OpenMarket/OpenMarket/Model/DataType/Currency.swift:
```diff
@@ -0,0 +1,11 @@
+//
+
+enum Currency: String, Codable {
+    case KRW = "KRW"
+    case USD = "USD"
```
`rawValue` 타입이 `String`이고, 케이스의 이름과 같으니 `rawValue`를 지정해주는 코드는 생략할 수 있을 것 같아요!

dnwhd0112 replied:
넵 생략하도록 하겠습니다!

TTOzzi commented on OpenMarket/OpenMarket/Model/DataType/PageInformation.swift:
```diff
@@ -0,0 +1,30 @@
+    let hasNext: Bool
+    let hasPrev: Bool
+    
+    private enum CodingKeys: String, CodingKey {
+        case pageNo = "page_no"
+        case itemsPerPage = "items_per_page"
+        case totalCount = "total_count"
+        case offset
+        case limit
+        case pages
+        case lastPage = "last_page"
+        case hasNext = "has_next"
+        case hasPrev = "has_prev"
+    }
```
선언하신 데이터 타입에서 `프로퍼티 이름` <-> `서버에서 내려주는 json 데이터의 키` 차이가 카멜 케이스냐, 스네이크 케이스냐의 차이 뿐이라면 별도의 `CodingKeys` 지정 없이 `JSONDecoder` 의 `keyDecodingStrategy` 를 활용해서 디코딩할 수도 있을 것 같습니다.
(이런 방법도 있다는 것을 알려드리는 것일 뿐 잘못됐다는 건 아니에요! 수정하실 필요는 없습니다 🙏 )

dnwhd0112 replied:
감사합니다. 한번 찾아보도록 할게요!

TTOzzi commented on OpenMarket/OpenMarket/Model/DataType/RegistrationParameter.swift:
```diff
@@ -0,0 +1,28 @@
+    let descriptions: String
+    let price: Double
+    let currency: Currency
+    let discountedPrice: Double = 0
+    let stock: Int = 0
```
여기만 따로 값을 지정해준 이유가 궁금해요 👀

dnwhd0112 replied:
default 값 설정때문에 따로 지정했습니다. 아직 이쪽은 실제 통신을 안해봐서 이 방법이 잘못됬을수도 있어요!

TTOzzi replied:
이건 다음 스텝에서 의도하신대로 값이 없을 때 기본값이 잘 들어가는지, 값이 있을 땐 서버에서 내려준 값을 사용하는지 확인해보아요 🙂

TTOzzi commented on OpenMarket/OpenMarket/Model/Network/Network.swift:
```diff
@@ -0,0 +1,30 @@
+    
+    func requestData(url: String,
+                     completeHandler: @escaping (Data?, URLResponse?, Error?) -> Void ) {
+        let configure = URLSessionConfiguration.default
+        let session = URLSession(configuration: configure)
```
매 요청마다 새로운 `URLSession`을 만들어야 하는 이유가 있을까요? 👀 
추가로 현재 구현은 `Network`의 로직이 `URLSession`과 강하게 결합되어 있습니다.
`URLSession`과 분리된 독립적인 테스트가 불가능할 것 같아요.
프로토콜 추상화와 의존성 주입을 활용해 `Network`와 `URLSession`간의 의존성을 줄여보면 좋을 것 같습니다!

dnwhd0112 replied:
없습니다 ㅎㅎ 사실 shared를 쓸까 생각도 했었는대 또치 말을 듣고 보니 분리를 해주는게 맞겠군요. 생성자로 Session을 받는 방식으로 구성을 해야될듯하네요. 프로토콜로 추상화를 사용하는 방법은... 잘 모르겠습니다.

westeastyear replied:
야곰닷넷에 나와있는 강의를 토대로 
```swift
typealias DataTaskComletionHandler = (Data?, URLResponse?, Error?) -> Void

protocol URLSessionProtocol {
    func dataTask(with url: URL, completionHandler: @escaping DataTaskComletionHandler) -> URLSessionDataTask
}

extension URLSession: URLSessionProtocol { }

------------

var session: URLSessionProtocol
    
    init(session: URLSessionProtocol) {
        self.session = session
    }
```
프로토콜을 만들고 추상화하여, 의존성을 주입해주는 방식으로 해보았습니다.

TTOzzi commented on OpenMarket/OpenMarket/Model/Network/Network.swift:
```diff
@@ -0,0 +1,30 @@
+        
+        let urlComponents = URLComponents(string: url)
+        
+        guard let requestURL = urlComponents?.url else { return }
```
인자로 넘겨받은 `String` 타입 `url`이 양식이 맞지 않는 등의 이유로 `URL`로 변환되지 않았을 때 함수를 종료해주고 있네요 🤔 
이 메서드를 사용하는 입장에선 어떤 이유로 네트워크 요청이 들어가지 않았는지 인지하기 어려울 것 같습니다.
예외 상황이 발생했을 때, 함수를 호출한 곳에서 어떤 예외 상황이 발생한 것인지 알 수 있도록 구현해주면 더 좋을 것 같아요.
`Network` 객체가 로직을 수행하면서 발생하는 예외 상황들을 커스텀 에러 타입으로 정의해서 넘겨준다거나 하는 방식으로요!

dnwhd0112 replied:
네 에러 타입을 만들고 이를 반환하도록 변경하겠습니다.
```swift
guard let requestURL = urlComponents?.url else {
            throw NetworkError.urlError
        }
```

TTOzzi commented on OpenMarket/OpenMarket/Model/Network/Network.swift:
```diff
@@ -0,0 +1,30 @@
+        let urlComponents = URLComponents(string: url)
+        
+        guard let requestURL = urlComponents?.url else { return }
+        let dataTask = session.dataTask(with: requestURL) { (data, response, error) in
+            completeHandler(data, response, error)
+        }
```
지금 구현은 `dataTask`를 통해 받은 결과를 그대로 넘겨주기만 하고 있어서, `Network` 객체의 역할이 거의 없는 것 같아요 🤔 
요청의 성공/실패도 `requestData` 메서드를 호출하는 곳 마다 구분해줘야 할 것 같구요.
`Network` 객체에서 http status code 확인,  응답으로 받은 data가 비어있는지 확인해서 성공/실패 구분 정도는 해줄 수 있지 않을까요? 👀

dnwhd0112 replied:
네 수정하도록 하겠습니다!

dnwhd0112 replied:
error 를 NetworkErro로 정의해주고 error 핸들러를 받도록 하여 이 핸들러에서 처리하도록 수정하였습니다.

TTOzzi commented on OpenMarket/OpenMarket/Model/Network/OpenMarketApi.swift:
```diff
@@ -0,0 +1,22 @@
+//  Created by 우롱차, Donnie on 2022/05/11.
+//
+
+enum OpenMarketApi {
```
API 요청에 대한 정보를 요렇게 분리해두신 것 정말 좋아요 😊
현재는 GET 요청으로 URL path와 queryItem 만을 활용하고 있어 문자열 조립 정도만 수행해주고 있지만, 이후 다른 방식의 요청을 사용해야 하거나 요청의 body로 JSON 데이터를 보내야 하는 상황이 온다면 `URLRequest` 를 만들어 반환해주는 것도 좋을 것 같네요.

TTOzzi commented on OpenMarket/OpenMarket/Model/Network/OpenMarketApi.swift:
```diff
@@ -0,0 +1,22 @@
+    case pageInformation(pageNo: Int, itemsPerPage: Int)
+    case productDetail(productNumber: Int)
+    
+    var string: String {
```
음 좀 더 명확한 이름으로 바꿔볼 수 있을까요?!

dnwhd0112 replied:
string 이부분 말씀이시죠? 고민해보겠습니다!

dnwhd0112 replied:
pathString으로 바꿨습니다.

TTOzzi commented on OpenMarket/OpenMarketJSONTests/OpenMarketJSONTests.swift:
```diff
@@ -0,0 +1,99 @@
+    
+    func test_products_JSON_테스트파일을_decoding해서_결과는_NotNil() {
+        // given
+        let promise = expectation(description: "비동기 메서드 테스트")
```
테스트에 실패하면, `description`과 함께 실패 로그가 남게 될 텐데 "비동기 메서드 테스트" 만으론 어떻게 실패한건지 구분하기 어려울 것 같아요.
어떤 로직을 만족하지 못해서 실패한 것인지 알 수 있도록 좀 더 자세하게 설정해주면 좋을 것 같습니다!

dnwhd0112 replied:
앗 description을 신경쓰지 못했군요. session time out으로 변경하면 될까요?

TTOzzi replied:
넵! 해당 `expectation`에 의해 테스트 케이스가 실패했다는 것은 요청에 대한 응답을 받지 못했다는 것이니 그걸 인지할 수 있는 텍스트면 뭐든 괜찮을 것 같아요.

TTOzzi commented on OpenMarket/OpenMarketJSONTests/OpenMarketJSONTests.swift:
```diff
@@ -0,0 +1,99 @@
+        let url = OpenMarketApi.pageInformation(pageNo: pageNo, itemsPerPage: itemsPerPage).string
+        
+        // when
+        mockNetwork.requestData(url: url) { data, response, error in
```
음.. 지금은 테스트를 돕기 위해 모의 객체를 사용하는 것이 아닌, 모의 객체의 로직 자체를 테스트하고 있는 것 같네요.
이 테스트가 성공한다고 해서, 실제로 프로덕트에 사용되는 `Network` 객체가 정상적으로 동작한다는 것을 보장할 수 있을까요? 👀 

`Network` 객체의 로직을 실제 네트워킹 없이 테스트하기 위한 모의 객체를 구현, 주입하고 `Network`의 로직에 대한 테스트가 수행되어야 할 것 같습니다.
지금 구현하신 로직에선 이해가 어려우실 수 있어요. `Network` 에서 드렸던 `URLSession`과의 의존성에 대한 코멘트와 함께 고민해보시면 좋을 것 같아요.

dnwhd0112 replied:
수정했습니다! 로직을 변경하여 session 및 datatask를 다르게 반환 받도록 변경했습니다.

TTOzzi commented on OpenMarket/OpenMarketJSONTests/OpenMarketJSONTests.swift:
```diff
@@ -0,0 +1,99 @@
+        
+        // when
+        network.requestData(url: url) { data, response, error in
+            guard error == nil else { return }
+            
+            let successsRange = 200..<300
+            
+            guard let statusCode = (response as? HTTPURLResponse)?.statusCode,
+                    successsRange.contains(statusCode) else { return }
+            
+            guard let data = data,
```
매 요청마다 공통으로 들어가는 이 성공 여부 검사 로직을 `Network`로 옮겨주면 좋을 것 같네요!

dnwhd0112 replied:
네 이부분은 Network로 옮긴뒤 오류 처리 핸들러를 또 추가하여서 실패시 오류처리 핸들러가 처리하도록 수정하였습니다.

TTOzzi commented on OpenMarket/OpenMarketJSONTests/OpenMarketJSONTests.swift:
```diff
@@ -0,0 +1,99 @@
+        wait(for: [promise], timeout: 10)
+    }
+    
+    func test_productDetail_decoding해서_결과는_NotNil() {
```
지금은 성공하는 경우에 대한 테스트만 작성되어있는데요.
로직이 실패하는 경우에도 그에 맞는 예외처리가 되어있는지에 대한 테스트도 중요합니다.
지금 구현하신 로직에선 잘못된 `URL`을 넘겨줬을 때의 처리를 테스트 해볼 수 있겠네요!
(`Network` 에서 성공, 실패 판단까지 하게 되면 더 많은 테스트를 해볼 수 있을 것 같아요.)

dnwhd0112 replied:
넵! Mock 객체를 수정하고 Network에서 에러가 발생시 테스트까지 마쳤습니다.

TTOzzi reviewed (changes requested):
안녕하세요 도니, 우롱차!
제가 놓친 부분이 있거나, 남겨드린 코멘트 중 이해가 어려운 부분이 있다면 언제든지 말씀해주세요 :pray:
고생하셨습니다 :bow:

### `requestData` 구현시 어디까지 `escaping Handler`에서 처리할것이고, 어디까지 함수 내부에서 처리를 한 뒤 `Handler`에게 넘겨줄지를 고민하였습니다.

코멘트로 남겨드리긴 했는데, 현재는 `Network` 객체의 역할이 `URLSession`의 `dataTask` 메서드를 단순히 래핑하는 정도인 것 같아요.
`Network` 객체의 역할, 구현 목적이 무엇인지 고민해보시면 좋을 것 같습니다!

### `URL` 주소를 관리하는 방법에 대해

:thumbsup:

### 네이밍 `Network`가 부적절한거 같은데 딱히 더 좋은 방안이 생각나지 않았습니다. 네이밍에 대한 많은 조언 부탁드립니다.🙇🏻‍♂️🙇🏻‍♂️

NetworkService, NetworkManager, NetworkProvider 등 다양한 네이밍을 사용하는데, 딱히 정해진 건 없어요!
네트워크 관련 로직을 수행하는 객체임을 명확하게 알 수 있다면 어떤 네이밍을 사용하던 문제 없다고 생각합니다.

### 테스트하는 `JSON`파일이 `main asset`에 들어가 있는게 좋아보이지 않아서 `test`파일에 위치해두어 `testBundle`로 접근하게 하였습니다. 저희는 괜찮다고 생각이 드는데, 또치의 생각은 어떠한가요?

너무 좋아요! 테스트를 위한 JSON 파일까지 함께 빌드에서 앱스토어에 올릴 필요는 없죠. 파일 타겟까지 너무 잘 나눠주셨습니다.

### `Session`으로부터 `response`를 받을때 모든 `data`를 저장하도록 객체를 설계하였는데, 필요한 `data`만 디코딩하여 저장하는것이 나은지 궁금하여 조언을 받고싶습니다.

사용하지 않는 값도 디코딩해서 갖고있어야 하는가 가 고민이신 건가요? :eyes:
사용하지 않는 값까지 굳이 가지고 있을 필요는 없을 것 같아요. 혹시나 나중에 필요해진다면 그 때 추가해도 되니까요!

dnwhd0112 committed:
♻️[refactor]: 1차 피드백 반영

Session 프로토콜 추상화 및 Network 의존성 주입, error 타입 생성, Network에서 DataTaskCompletionHandler 함수 코드 변경

westeastyear committed:
🚚[move]: URLSessionProtocol 파일 생성 및 코드 이동

westeastyear committed:
♻️[refactor]: 테스트를 위한 코드 수정

dnwhd0112 committed:
♻️[refactor]: 테스트 코드 수정 및 StubURLSessionDataTask에 함수 추가

westeastyear committed:
Merge branch 'STEP1-2'

TTOzzi commented on OpenMarket/OpenMarket/Model/DataType/Image.swift:
```diff
@@ -0,0 +1,24 @@
+
+import Foundation
+
+struct Image: Codable {
```
구현된 로직에서 디코드만 하고 있는 것 같아 `Codable` 보단 `Decodable`을 사용하는게 좋을 것 같아요.

추가로 `Image`와 같이 너무 일반적인 네이밍은 지금 당장은 충돌이 일어나지 않더라도 나중에 애플에서 제공하는 타입과 이름이 겹쳐 충돌이 발생할 수 있습니다.
(UIKit을 사용하는 프로젝트라 문제 없긴 하지만.. SwiftUI의 Image 타입과 이름이 겹쳐요!)

dnwhd0112 replied:
ProductRegistration에서 Image를 인코딩 하는 경우가 있을꺼 같아 Codable로 유지하도록 하겠습니다. 네이밍은  ProductImage 로 바꿧습니다!

TTOzzi commented on OpenMarket/OpenMarket/Model/Network/Network.swift:
```diff
@@ -0,0 +1,61 @@
+
+protocol NetworkAble {
+    func requestData(url: String,
+                     completeHandler: @escaping DataTaskCompletionHandler,
```
typealias를 맞춰주기 위해 불필요한 인자(Error)를 넘기는 것이 괜찮은지

dnwhd0112 replied:
error를 안넘기도록 수정하였습니다.

TTOzzi commented on OpenMarket/OpenMarket/Model/Network/Network.swift:
```diff
@@ -0,0 +1,61 @@
+        static let successRange = 200..<300
+    }
+    
+    var session: URLSessionProtocol
```
외부에서 접근하는 케이스가 있는지, 변경될 여지가 있는지

dnwhd0112 replied:
private let 으로 변경하였습니다.

TTOzzi commented on OpenMarket/OpenMarket/Model/Network/Network.swift:
```diff
@@ -0,0 +1,61 @@
+            return
+        }
+        
+        let dataTask = session.dataTask(with: requestURL) { (data, response, error) in
```
보냈던 요청을 취소할 수 있도록 구현을 확장해보는 것
(요건 굳이 반영 안 하셔도 됩니다 🙏 )

dnwhd0112 replied:
return dataTask를 하도록 수정하였습니다.

TTOzzi commented on OpenMarket/OpenMarketJSONTests/MockNetwork.swift:
```diff
@@ -0,0 +1,42 @@
+
+import Foundation
+
+//final class MockNetwork: NetworkAble {
```
불필요한 코드

TTOzzi commented on OpenMarket/OpenMarketJSONTests/OpenMarketJSONTests.swift:
```diff
@@ -0,0 +1,154 @@
+        let mockNetwork = Network(session: stubUrlSession)
+        
+        //when
+        mockNetwork.requestData(url: OpenMarketApi.pageInformation(pageNo: 1, itemsPerPage: 10).string)
+        { (data: Data?, response: URLResponse?, error: Error?) -> Void in
```
컨벤션 + 타입 선언, Void 반환 생략 가능한지

dnwhd0112 replied:
수정 완료했습니다!

TTOzzi commented on OpenMarket/OpenMarketJSONTests/OpenMarketJSONTests.swift:
```diff
@@ -0,0 +1,154 @@
+                  let pageInformation = try? JSONDecoder().decode(PageInformation.self, from: data) else { return }
+            
+            //then
+            XCTAssertNotNil(pageInformation)
+            XCTAssertEqual(pageInformation.pages.first?.name, "Test Product" )
```
`pageInformation.pages.first?.name`만 같다고 정상적인 데이터가 받아졌다는 것을 보장할 수 있는지

dnwhd0112 replied:
테스트 파일로 한경우는 그렇게 진행하였고 서버로부터 통신해서 데이터 받는경우는 데이터가 달라질수 있기때문에 Nil 확인으로 진행하였습니다.

TTOzzi commented on OpenMarket/OpenMarketJSONTests/OpenMarketJSONTests.swift:
```diff
@@ -0,0 +1,154 @@
+            promise.fulfill()
+        }
+    errorHandler: { (error: Error) -> Void in
+        print(error)
+        XCTFail()
```
print 보단 XCTFail의 message 활용

dnwhd0112 replied:
수정완료했습니다.

TTOzzi commented on OpenMarket/OpenMarketJSONTests/OpenMarketJSONTests.swift:
```diff
@@ -0,0 +1,154 @@
+        wait(for: [promise], timeout: 10)
+    }
+    
+    func test_mockNetwork객체를_sessionError를_발생시켜서_확인() {
```
나머지 3개의 예외 상황에 대한 테스트

TTOzzi commented on OpenMarket/OpenMarketJSONTests/OpenMarketJSONTests.swift:
```diff
@@ -0,0 +1,154 @@
+        
+        //when
+        mockNetwork.requestData(url: OpenMarketApi.pageInformation(pageNo: 1, itemsPerPage: 10).string) { (data: Data?, response: URLResponse?, error: Error?) -> Void in
+            guard let data = data,
+                  let pageInformation = try? JSONDecoder().decode(PageInformation.self, from: data) else { return }
+            
+            //then
+            XCTAssertNotNil(pageInformation)
+            XCTAssertEqual(pageInformation.pages.first?.name, "Test Product" )
```
실패하는 경우에 대한 테스트에선 불필요한 코드

dnwhd0112 replied:
수정완료했습니다.

TTOzzi reviewed (approved):
전체적인 피드백은 구두로 전달드려서 잊어버리시지 않도록 기록으로 남기는 정도로만 간단히 코멘트 남겼어요.
(대부분 자세히 설명드렸다 생각해서 짤막짤막하게 적었는데 양해 부탁드립니다.. 🙏 )
추가로 코드 한 번 더 보면서 말씀 못 드렸던 부분도 코멘트 남겨뒀어요.
말로 하다보니 이것저것 많이 말씀드리긴 했는데 두 분이 다음 스텝 진행하시면서 납득이 되는 피드백만 적당히 골라서 반영하시면 좋을 것 같습니다.
고생하셨습니다~! 🙇
